File: Spark/Ejercicios/chapter3/ejemplo1.scala
package main.scala.chapter3.ejemplo1

// Imports
import org.apache.spark.SparkContext
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

object ejemplo {
    def main(args: Array[String]) {
        val sc = SparkContext.getOrCreate();
        // The virgin RDD: Ininteligible y confuso, bajo nivel

        val dataRDD = sc.parallelize(Array(("Brooke",20),("Denny",31),("Jules",30),("TD",35),("Brooke",25)))
        val agesRDD = dataRDD.map(x => (x._1, (x._2, 1))).reduceByKey((x, y) => (x._1+y._1,x._2+y._2)).map(x => (x._1, (x._2._1/x._2._2)))

        agesRDD.take(5) // res10: Array[(String, Int)] = Array((Denny,31), (TD,35), (Brooke,22.5), (Jules,30))

        // The chad DF: Entendible, alto nivel
        
        // Creamos la SparkSession
        val spark = SparkSession.builder.appName("ejemplo1").getOrCreate()

        val data_df = spark.createDataFrame(Seq(("Brooke",20), ("Brooke", 25), ("Denny", 31), ("Jules", 30), ("TD", 35))).toDF("name","age")
        data_df.printSchema()
        val avg_df = data_df.groupBy("name").agg(avg("age"))
        avg_df.show() // Muestra una tabla muy bonica con los mismos datos de arriba
    }
}

File: Spark/Ejercicios/chapter3/ejemplo1.py
#Imports
import sys

from pyspark.sql import SparkSession
from pyspark import SparkContext, SparkConf
from pyspark.sql.functions import avg

# The virgin RDD: Ininteligible y confuso, bajo nivel

sc = SparkContext()
dataRDD = sc.parallelize([("Brooke",20),("Denny",31),("Jules",30),("TD",35),("Brooke",25)])
agesRDD = dataRDD.map(lambda x: (x[0], (x[1], 1))).reduceByKey(lambda x, y: (x[0]+y[0],x[1]+y[1])).map(lambda x: (x[0], (x[1][0]/x[1][1])))

# The chad DF: Entendible, alto nivel

spark = SparkSession.builder.appName("Ejemplo1").getOrCreate()

data_df = spark.createDataFrame([("Brooke",20),("Denny",31),("Jules",30),("TD",35),("Brooke",25)], ["name", "age"])
avg_df = data_df.groupBy("name").agg(avg("age"))
avg_df.show()